sklearn/naive_bayes.py

```diff
@@ -958,3 +998,256 @@ def _joint_log_likelihood(self, X):
+        """Incremental fit on a batch of samples.
+
+        This method is expected to be called several times consecutively
+        on different chunks of a dataset so as to implement out-of-core
+        or online learning.
+
+        This is especially useful when the whole dataset is too big to fit in
+        memory at once.
+
+        This method has some performance overhead hence it is better to call
+        partial_fit on chunks of data that are as large as possible
+        (as long as fitting in the memory budget) to hide the overhead.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = [n_samples, n_features]
+            Training vectors, where n_samples is the number of samples and
+            n_features is the number of features. Here, each feature of X is
+            assumed to be from a different categorical distribution.
+
+        y : array-like, shape = [n_samples]
+            Target values.
+
+        classes : array-like, shape = [n_classes] (default=None)
+            List of all the classes that can possibly appear in the y vector.
+
+            Must be provided at the first call to partial_fit, can be omitted
+            in subsequent calls.
+
+        Returns
+        -------
+        self : object
+        """
+        return super(CategoricalNB, self).partial_fit(X, y, classes,
+                                                      sample_weight=None)
+
+    def _check_input(self, X, Y=None):
+        if Y is None:
+            if _old_numpy:
+                X = check_array(X, accept_sparse=False)
+                if not np.isfinite(X).all():
+                    raise ValueError("Input contains NaN or infinity.")
+                return np.asarray(X, dtype=np.int64)
+            else:
+                return check_array(X, accept_sparse=False)
+        else:
+            if _old_numpy:
+                X, Y = check_X_y(X, Y, accept_sparse=False)
+                if not np.isfinite(X).all():
+                    raise ValueError("Input contains NaN or infinity.")
+                X = np.asarray(X, dtype=np.int64)
+                return X, Y
+            else:
+                return check_X_y(X, Y, accept_sparse=False)
+
+    def _init_counters(self, n_effective_classes, n_features):
+        self.class_count_ = np.zeros(n_effective_classes, dtype=np.float64)
+        self.category_count_ = [np.zeros((self.class_count_.shape[0], 0))
+                                for _ in range(n_features)]
+        self.feature_cat_index_mapping_ = [{} for _ in range(n_features)]
+
+    def _count(self, X, Y):
+        if np.any((X.data if issparse(X) else X) < 0):
+            raise ValueError("Input X must be non-negative")
+        self.class_count_ += Y.sum(axis=0)
+        for i in range(self.n_features_):
+            X_feature = X[:, i]
+            categories = np.unique(X_feature)
+            self._update_mapping(self.feature_cat_index_mapping_[i],
```

Comment:
Hm, I am not sure, whether I understand you correctly here. Correct, we give no constraints on the labels, as long as they are numeric. However, we just store a mapping from a category to an index and the count of the respective category.
Do you mean we should require that all categories of the training data `X` are integers from 0 to n_cats - 1, in order to simplify the mechanics of CatNB?
I think even if we could assume this, the mechanics would still be similar (if not the same). Where would you simplify?:
- Iterate over all features
- get all categories for respective feature
- in case partial fit is used -> check if any new categories are found
- count

Reply:
@jnothman, I think we also need this here since `OrdinalEncoder` has no `handle_unknown` parameter yet. This is why dictionaries are used here and some parts of `OrdinalEncoder`'s functionality is repeated. Handling unknown categories during prediction is an important feature for this estimator and it would be rather useless otherwise. Do you have a better way to achieve this?

Reply:
you can still handle unknowns by just identifying the max value in each column if you require that they range from 0 to max.

Reply:
@jnothman I am not sure what to do now. Should I refactor the code and assume categorical data stored as integers from 0 to n_cats -1, or leave it as it is?